core/src/main/java/org/elasticsearch/transport/TransportChannelResponseHandler.java

```diff
@@ -67,15 +67,21 @@ protected TransportChannelResponseHandler(ESLogger logger, TransportChannel chan
 
     @Override
     public void handleResponse(T response) {
+        onFinished();
         try {
             channel.sendResponse(response);
         } catch (IOException e) {
             handleException(new TransportException(e));
         }
     }
 
+    protected void onFinished() {}
+
+    ;
+
     @Override
     public void handleException(TransportException exp) {
+        onFinished();
```

Comment:
fair enough and good catch. I removed it in favor of an override of the explicit methods. In the specific case I was using it's safe to run the onFinish logic twice.